src/graphql/replace.js

```diff
@@ -0,0 +1,73 @@
+/*
+ * vim: set tw=78 et ts=4 sw=4 si fileencoding=utf-8:
+ */
+const _ = require("lodash");
+const FIND_HISTORY_INPUT = /history(.*){/g;
+const FIND_FILTER_INPUT = /filter:\s?"(.*)"/g;
+
+/**
+ * An utility for replacing placeholder
+ * variables defined in a widesky query.
+ * e.g. $from and $to variables of the history
+ * node.
+ **/
+class Replace {
+
+    static outerBraces(query) {
+        let trimmedQ = _.trimStart(query);
+
+        if (!_.startsWith(trimmedQ, '{')) {
+            query = `{ ${query} }`;
+        }
+
+        return query;
+    }
+
+    static filterVar(
+        query,
+        varname,
+        varval) {
+        varname = _.trimStart(varname);
+        if (!_.startsWith(varname, '$')) {
+            varname = `\$${varname}`;
+        }
+
+        // Forward slash the $ character for regexpr,
+        varname = "\\" + varname;
+        const pattern = new RegExp(varname, 'g');
+
+        return query.replace(pattern, varval);
+    }
```

Comment:
I could see this tripping up when called multiple times because the result of one replacement may look like a placeholder in subsequent calls depending on the inputs given.  It also doesn't appear to handle trailing characters…

```javascript
var query = "… blah some query $to $toucan …";
query = Replace.filterVar(query, "to", "VAL");
// query now: "… blah some query VAL VALucan …"
```

Better, might be to use a [function as a replacement value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#Specifying_a_function_as_a_parameter) with a more generic regex such as `(\$[a-zA-Z_][a-zA-Z0-9_]*|\$\{[a-zA-Z_][a-zA-Z0-9_]*\})` which will pick up _all_ place-holders, then the function can pick out the value from a given object.

Our API server's "template" algorithm (used in authorisation logic) would be worth looking at as it avoids this problem and has been in production for some time now.

Reply:
I've raised BA-640